File: src/services/command_service.py
from __future__ import annotations

import asyncio
import shlex
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable, Dict, List, Optional, Sequence, Tuple

from src.utils.platform import get_default_shell, PlatformType, detect_platform


# ---------------------------------------------------------------------------
# Event Emitter (minimal, Node-like)
# ---------------------------------------------------------------------------

class EventEmitter:
    def __init__(self) -> None:
        self._listeners: Dict[str, List[Callable[..., None]]] = {}

    def on(self, event: str, callback: Callable[..., None]) -> None:
        self._listeners.setdefault(event, []).append(callback)

    def off(self, event: str, callback: Callable[..., None]) -> None:
        if event in self._listeners:
            try:
                self._listeners[event].remove(callback)
            except ValueError:
                pass

    def emit(self, event: str, *args, **kwargs) -> None:
        for cb in list(self._listeners.get(event, [])):
            try:
                cb(*args, **kwargs)
            except Exception:
                # Keep emitter resilient (like Node's default behavior)
                pass


# ---------------------------------------------------------------------------
# Types (mirror TS interfaces/enums)
# ---------------------------------------------------------------------------

class CommandSecurityLevel(str, Enum):
    SAFE = "safe"
    REQUIRES_APPROVAL = "requires_approval"
    FORBIDDEN = "forbidden"


@dataclass
class CommandWhitelistEntry:
    command: str
    security_level: CommandSecurityLevel
    allowed_args: Optional[List[str]] = None  # NOTE: TS allowed RegExp; see note below
    description: Optional[str] = None


@dataclass
class CommandResult:
    stdout: str
    stderr: str


@dataclass
class PendingCommand:
    id: str
    command: str
    args: List[str]
    requested_at: float  # epoch seconds
    requested_by: Optional[str]
    resolve: Callable[[CommandResult], None]
    reject: Callable[[Exception], None]


# ---------------------------------------------------------------------------
# CommandService
# ---------------------------------------------------------------------------

class CommandService(EventEmitter):
    def __init__(self, shell: Optional[str] = None, default_timeout: int = 30_000) -> None:
        """
        :param shell: Shell to use for commands (default: auto-detected per platform)
        :param default_timeout: Default timeout (ms) for command execution
        """
        super().__init__()
        self.shell: str = shell or get_default_shell()
        self.whitelist: Dict[str, CommandWhitelistEntry] = {}
        self.pending_commands: Dict[str, PendingCommand] = {}
        self.default_timeout: int = default_timeout

        # Initialize with platform-specific commands
        self._initialize_default_whitelist()

    # ---------------------------------------------------------------------
    # Public getters
    # ---------------------------------------------------------------------

    def get_shell(self) -> str:
        return self.shell

    def get_whitelist(self) -> List[CommandWhitelistEntry]:
        return list(self.whitelist.values())

    def get_pending_commands(self) -> List[PendingCommand]:
        return list(self.pending_commands.values())

    # ---------------------------------------------------------------------
    # Whitelist management
    # ---------------------------------------------------------------------

    def add_to_whitelist(self, entry: CommandWhitelistEntry) -> None:
        self.whitelist[entry.command] = entry

    def remove_from_whitelist(self, command: str) -> None:
        self.whitelist.pop(command, None)

    def update_security_level(self, command: str, security_level: CommandSecurityLevel) -> None:
        entry = self.whitelist.get(command)
        if entry:
            entry.security_level = security_level
            self.whitelist[command] = entry

    # ---------------------------------------------------------------------
    # Internal: whitelist init
    # ---------------------------------------------------------------------

    def _initialize_default_whitelist(self) -> None:
        # Late import to avoid circular dependency
        from src.utils.command_whitelist import get_platform_specific_commands

        platform_commands = get_platform_specific_commands()
        for entry in platform_commands:
            # normalize to dict key as base command (same as TS using path.basename)
            base = entry.command.split("\\")[-1].split("/")[-1]
            self.whitelist[base] = entry

    # ---------------------------------------------------------------------
    # Validation
    # ---------------------------------------------------------------------

    def _validate_command(self, command: str, args: Sequence[str]) -> Optional[CommandSecurityLevel]:
        """
        Validate if a command and its arguments are allowed.
        Returns the command's security level or None if not whitelisted.
        """
        base_command = command.split("\\")[-1].split("/")[-1]
        entry = self.whitelist.get(base_command)
        if not entry:
            return None

        if entry.security_level == CommandSecurityLevel.FORBIDDEN:
            return CommandSecurityLevel.FORBIDDEN

        # NOTE: TS supports allowedArgs as (string | RegExp) per arg-position.
        # This direct port treats allowed_args as exact string matches by position.
        # If you need regex, we can extend this to accept compiled patterns.
        if entry.allowed_args:
            if len(args) > len(entry.allowed_args):
                return CommandSecurityLevel.REQUIRES_APPROVAL

            for idx, arg in enumerate(args):
                pat = entry.allowed_args[idx] if idx < len(entry.allowed_args) else None
                if pat is None:
                    return CommandSecurityLevel.REQUIRES_APPROVAL
                # Exact match only (line-for-line conservative port)
                if arg != pat:
                    return CommandSecurityLevel.REQUIRES_APPROVAL

        return entry.security_level

    # ---------------------------------------------------------------------
    # Execution
    # ---------------------------------------------------------------------

    async def execute_command(
        self,
        command: str,
        args: Optional[List[str]] = None,
        *,
        timeout: Optional[int] = None,
        requested_by: Optional[str] = None,
    ) -> CommandResult:
        """
        Execute a shell command.
        """
        args = args or []
        security_level = self._validate_command(command, args)

        if security_level is None:
            raise RuntimeError(f"Command not whitelisted: {command}")

        if security_level == CommandSecurityLevel.FORBIDDEN:
            raise RuntimeError(f"Command is forbidden: {command}")

        if security_level == CommandSecurityLevel.REQUIRES_APPROVAL:
            return await self._queue_command_for_approval(command, args, requested_by)

        # SAFE: execute immediately
        return await self._run_now(command, args, timeout=timeout)

    async def approve_command(self, command_id: str) -> CommandResult:
        pending = self.pending_commands.get(command_id)
        if not pending:
            raise RuntimeError(f"No pending command with ID: {command_id}")

        try:
            result = await self._run_now(pending.command, pending.args, timeout=None)
            # Remove and emit
            self.pending_commands.pop(command_id, None)
            self.emit("command:approved", {"commandId": command_id, "stdout": result.stdout, "stderr": result.stderr})
            pending.resolve(result)
            return result
        except Exception as e:
            self.pending_commands.pop(command_id, None)
            self.emit("command:failed", {"commandId": command_id, "error": e})
            pending.reject(e)
            raise

    def deny_command(self, command_id: str, reason: str = "Command denied") -> None:
        pending = self.pending_commands.get(command_id)
        if not pending:
            raise RuntimeError(f"No pending command with ID: {command_id}")

        self.pending_commands.pop(command_id, None)
        self.emit("command:denied", {"commandId": command_id, "reason": reason})
        pending.reject(RuntimeError(reason))

    # ---------------------------------------------------------------------
    # Non-blocking queue (fire-and-forget)
    # ---------------------------------------------------------------------

    def queue_command_for_approval_non_blocking(
        self, command: str, args: Optional[List[str]] = None, requested_by: Optional[str] = None
    ) -> str:
        args = args or []
        cmd_id = str(uuid.uuid4())
        pending = PendingCommand(
            id=cmd_id,
            command=command,
            args=list(args),
            requested_at=asyncio.get_event_loop().time(),
            requested_by=requested_by,
            resolve=lambda _res: None,
            reject=lambda _err: None,
        )
        self.pending_commands[cmd_id] = pending
        self.emit("command:pending", pending)

        # 5s warning timeout to detect UI approval issues
        def _timeout_check() -> None:
            if cmd_id in self.pending_commands:
                self.emit(
                    "command:approval_timeout",
                    {
                        "commandId": cmd_id,
                        "message": (
                            "Command approval timed out. If you approved this command in the UI, "
                            "please use get_pending_commands and approve_command to complete the process."
                        ),
                    },
                )

        loop = asyncio.get_event_loop()
        loop.call_later(5.0, _timeout_check)

        return cmd_id

    # ---------------------------------------------------------------------
    # Internal: queue with awaitable result
    # ---------------------------------------------------------------------

    async def _queue_command_for_approval(
        self, command: str, args: Optional[List[str]], requested_by: Optional[str]
    ) -> CommandResult:
        args = args or []
        loop = asyncio.get_event_loop()
        fut: asyncio.Future[CommandResult] = loop.create_future()
        cmd_id = str(uuid.uuid4())

        def resolve(res: CommandResult) -> None:
            if not fut.done():
                fut.set_result(res)

        def reject(err: Exception) -> None:
            if not fut.done():
                fut.set_exception(err)

        pending = PendingCommand(
            id=cmd_id,
            command=command,
            args=list(args),
            requested_at=loop.time(),
            requested_by=requested_by,
            resolve=resolve,
            reject=reject,
        )

        self.pending_commands[cmd_id] = pending
        self.emit("command:pending", pending)

        # 5s warning timeout (UI signal)
        def _timeout_check() -> None:
            if cmd_id in self.pending_commands:
                self.emit(
                    "command:approval_timeout",
                    {
                        "commandId": cmd_id,
                        "message": (
                            "Command approval timed out. If you approved this command in the UI, "
                            "please use get_pending_commands and approve_command to complete the process."
                        ),
                    },
                )

        loop.call_later(5.0, _timeout_check)
        return await fut

    # ---------------------------------------------------------------------
    # Internal: run command now (using the configured shell)
    # ---------------------------------------------------------------------

    async def _run_now(self, command: str, args: Sequence[str], timeout: Optional[int]) -> CommandResult:
        """
        Emulates Node's execFile(command, args, { shell }) by invoking the configured shell
        and passing the whole command line as a single string. Handles Windows vs POSIX.
        """
        platform = detect_platform()
        cmdline = " ".join([shlex.quote(command), *(shlex.quote(a) for a in args)])

        if platform == PlatformType.WINDOWS:
            # cmd.exe /c "<cmdline>"
            shell_cmd = [self.shell, "/c", cmdline]
        else:
            # /bin/bash -c "<cmdline>"  (or zsh, etc.)
            shell_cmd = [self.shell, "-c", cmdline]

        # Convert ms → seconds for asyncio timeout
        timeout_sec = (timeout if timeout is not None else self.default_timeout) / 1000.0

        proc = await asyncio.create_subprocess_exec(
            *shell_cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        try:
            stdout_b, stderr_b = await asyncio.wait_for(proc.communicate(), timeout=timeout_sec)
        except asyncio.TimeoutError:
            with contextlib.suppress(ProcessLookupError):
                proc.kill()
            raise RuntimeError("Command execution failed: timeout")

        stdout = stdout_b.decode(errors="replace") if stdout_b else ""
        stderr = stderr_b.decode(errors="replace") if stderr_b else ""
        return CommandResult(stdout=stdout, stderr=stderr)

File: src/services/command-service.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import { randomUUID } from 'crypto';
import { EventEmitter } from 'events';
import * as path from 'path';
import { getDefaultShell } from '../utils/platform-utils.js';
import { getPlatformSpecificCommands } from '../utils/command-whitelist-utils.js';

const execFileAsync = promisify(execFile);

/**
 * Command security level classification
 */
export enum CommandSecurityLevel {
  /** Safe commands that can be executed without approval */
  SAFE = 'safe',
  /** Commands that require approval before execution */
  REQUIRES_APPROVAL = 'requires_approval',
  /** Commands that are explicitly forbidden */
  FORBIDDEN = 'forbidden'
}

/**
 * Command whitelist entry
 */
export interface CommandWhitelistEntry {
  /** The command path or name */
  command: string;
  /** Security level of the command */
  securityLevel: CommandSecurityLevel;
  /** Allowed arguments (string for exact match, RegExp for pattern match) */
  allowedArgs?: Array<string | RegExp>;
  /** Description of the command for documentation */
  description?: string;
}

/**
 * Pending command awaiting approval
 */
export interface PendingCommand {
  /** Unique ID for the command */
  id: string;
  /** The command to execute */
  command: string;
  /** Arguments for the command */
  args: string[];
  /** When the command was requested */
  requestedAt: Date;
  /** Who requested the command */
  requestedBy?: string;
  /** Resolve function to call when approved */
  resolve: (value: { stdout: string; stderr: string }) => void;
  /** Reject function to call when denied */
  reject: (reason: Error) => void;
}

/**
 * Result of command execution
 */
export interface CommandResult {
  /** Standard output from the command */
  stdout: string;
  /** Standard error from the command */
  stderr: string;
}

/**
 * Command service configuration options
 */
export interface CommandServiceOptions {
  /** Optional shell path to use when shell execution is enabled */
  shell?: string;
  /** Whether to execute commands through a shell (default: false) */
  useShell?: boolean;
  /** Default timeout for command execution in milliseconds (default: 30000) */
  defaultTimeout?: number;
}

/**
 * Service for securely executing shell commands
 */
export class CommandService extends EventEmitter {
  /** Shell to use for commands */
  private shell: string;
  /** Whether shell parsing is enabled */
  private useShell: boolean;
  /** Command whitelist */
  private whitelist: Map<string, CommandWhitelistEntry>;
  /** Pending commands awaiting approval */
  private pendingCommands: Map<string, PendingCommand>;
  /** Default timeout for command execution in milliseconds */
  private defaultTimeout: number;

  /**
   * Create a new CommandService
   * @param options Command service configuration options
   */
  constructor(options: CommandServiceOptions = {}) {
    super();
    this.shell = options.shell || getDefaultShell();
    this.useShell = options.useShell ?? false;
    this.whitelist = new Map();
    this.pendingCommands = new Map();
    this.defaultTimeout = options.defaultTimeout ?? 30000;

    // Initialize with platform-specific commands
    this.initializeDefaultWhitelist();
  }

  /**
   * Get the current shell being used
   * @returns The shell path
   */
  public getShell(): string {
    return this.shell;
  }

  /**
   * Whether command execution uses a shell for parsing
   * @returns True if shell execution is enabled
   */
  public isShellEnabled(): boolean {
    return this.useShell;
  }

  /**
   * Initialize the default command whitelist based on the current platform
   */
  private initializeDefaultWhitelist(): void {
    // Get platform-specific commands
    const platformCommands = getPlatformSpecificCommands();
    
    // Add all commands to the whitelist
    platformCommands.forEach(entry => {
      this.whitelist.set(entry.command, entry);
    });
  }

  /**
   * Add a command to the whitelist
   * @param entry The command whitelist entry
   */
  public addToWhitelist(entry: CommandWhitelistEntry): void {
    this.whitelist.set(entry.command, entry);
  }

  /**
   * Remove a command from the whitelist
   * @param command The command to remove
   */
  public removeFromWhitelist(command: string): void {
    this.whitelist.delete(command);
  }

  /**
   * Update a command's security level
   * @param command The command to update
   * @param securityLevel The new security level
   */
  public updateSecurityLevel(command: string, securityLevel: CommandSecurityLevel): void {
    const entry = this.whitelist.get(command);
    if (entry) {
      entry.securityLevel = securityLevel;
      this.whitelist.set(command, entry);
    }
  }

  /**
   * Get all whitelisted commands
   * @returns Array of command whitelist entries
   */
  public getWhitelist(): CommandWhitelistEntry[] {
    return Array.from(this.whitelist.values());
  }

  /**
   * Get all pending commands awaiting approval
   * @returns Array of pending commands
   */
  public getPendingCommands(): PendingCommand[] {
    return Array.from(this.pendingCommands.values());
  }

  /**
   * Validate if a command and its arguments are allowed
   * @param command The command to validate
   * @param args The command arguments
   * @returns The security level of the command or null if not whitelisted
   */
  private validateCommand(command: string, args: string[]): CommandSecurityLevel | null {
    // Extract the base command (without path) using path.basename
    const baseCommand = path.basename(command);
    
    // Check if the command is in the whitelist
    const entry = this.whitelist.get(baseCommand);
    if (!entry) {
      return null;
    }

    // If the command is forbidden, return immediately
    if (entry.securityLevel === CommandSecurityLevel.FORBIDDEN) {
      return CommandSecurityLevel.FORBIDDEN;
    }

    // If there are allowed arguments defined, validate them
    if (entry.allowedArgs && entry.allowedArgs.length > 0) {
      // Check if all arguments are allowed
      const allArgsValid = args.every((arg, index) => {
        // If we have more args than allowed patterns, reject
        if (index >= (entry.allowedArgs?.length || 0)) {
          return false;
        }

        const pattern = entry.allowedArgs?.[index];
        if (!pattern) {
          return false;
        }

        // Check if the argument matches the pattern
        if (typeof pattern === 'string') {
          return arg === pattern;
        } else {
          return pattern.test(arg);
        }
      });

      if (!allArgsValid) {
        return CommandSecurityLevel.REQUIRES_APPROVAL;
      }
    }

    return entry.securityLevel;
  }

  /**
   * Execute a shell command
   * @param command The command to execute
   * @param args Command arguments
   * @param options Additional options
   * @returns Promise resolving to command output
   */
  public async executeCommand(
    command: string,
    args: string[] = [],
    options: {
      timeout?: number;
      requestedBy?: string;
    } = {}
  ): Promise<CommandResult> {
    const securityLevel = this.validateCommand(command, args);

    // If command is not whitelisted, reject
    if (securityLevel === null) {
      throw new Error(`Command not whitelisted: ${command}`);
    }

    // If command is forbidden, reject
    if (securityLevel === CommandSecurityLevel.FORBIDDEN) {
      throw new Error(`Command is forbidden: ${command}`);
    }

    // If command requires approval, add to pending queue
    if (securityLevel === CommandSecurityLevel.REQUIRES_APPROVAL) {
      return this.queueCommandForApproval(command, args, options.requestedBy);
    }

    // For safe commands, execute immediately
    try {
      const timeout = options.timeout || this.defaultTimeout;
      const { stdout, stderr } = await execFileAsync(command, args, {
        timeout,
        shell: this.useShell ? this.shell : false
      });

      return { stdout, stderr };
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`Command execution failed: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Queue a command for approval
   * @param command The command to queue
   * @param args Command arguments
   * @param requestedBy Who requested the command
   * @returns Promise resolving when command is approved and executed
   */
  private queueCommandForApproval(
    command: string,
    args: string[] = [],
    requestedBy?: string
  ): Promise<CommandResult> {
    return new Promise((resolve, reject) => {
      const id = randomUUID();
      const pendingCommand: PendingCommand = {
        id,
        command,
        args,
        requestedAt: new Date(),
        requestedBy,
        resolve: (result: CommandResult) => resolve(result),
        reject: (error: Error) => reject(error)
      };

      this.pendingCommands.set(id, pendingCommand);
      
      // Emit event for pending command
      this.emit('command:pending', pendingCommand);
      
      // Set a timeout to check if the command is still pending after a while
      // This helps detect if the UI approval didn't properly trigger the approveCommand method
      setTimeout(() => {
        // If the command is still pending after the timeout
        if (this.pendingCommands.has(id)) {
          // Emit a warning event that can be handled by the client
          this.emit('command:approval_timeout', {
            commandId: id,
            message: 'Command approval timed out. If you approved this command in the UI, please use get_pending_commands and approve_command to complete the process.'
          });
        }
      }, 5000); // 5 second timeout to detect UI approval issues
    });
  }

  /**
   * Queue a command for approval without waiting for the Promise to resolve
   * @param command The command to queue
   * @param args Command arguments
   * @param requestedBy Who requested the command
   * @returns The ID of the queued command
   */
  public queueCommandForApprovalNonBlocking(
    command: string,
    args: string[] = [],
    requestedBy?: string
  ): string {
    const id = randomUUID();
    const pendingCommand: PendingCommand = {
      id,
      command,
      args,
      requestedAt: new Date(),
      requestedBy,
      resolve: () => {}, // No-op resolve function
      reject: () => {}   // No-op reject function
    };

    this.pendingCommands.set(id, pendingCommand);
    
    // Emit event for pending command
    this.emit('command:pending', pendingCommand);
    
    // Set a timeout to check if the command is still pending after a while
    setTimeout(() => {
      // If the command is still pending after the timeout
      if (this.pendingCommands.has(id)) {
        // Emit a warning event that can be handled by the client
        this.emit('command:approval_timeout', {
          commandId: id,
          message: 'Command approval timed out. If you approved this command in the UI, please use get_pending_commands and approve_command to complete the process.'
        });
      }
    }, 5000); // 5 second timeout to detect UI approval issues
    
    return id;
  }

  /**
   * Approve a pending command
   * @param commandId ID of the command to approve
   * @returns Promise resolving to command output
   */
  public async approveCommand(commandId: string): Promise<CommandResult> {
    const pendingCommand = this.pendingCommands.get(commandId);
    if (!pendingCommand) {
      throw new Error(`No pending command with ID: ${commandId}`);
    }

    try {
      const { stdout, stderr } = await execFileAsync(
        pendingCommand.command,
        pendingCommand.args,
        { shell: this.useShell ? this.shell : false }
      );

      // Remove from pending queue
      this.pendingCommands.delete(commandId);
      
      // Emit event for approved command
      this.emit('command:approved', { commandId, stdout, stderr });
      
      // Resolve the original promise
      pendingCommand.resolve({ stdout, stderr });
      
      return { stdout, stderr };
    } catch (error) {
      // Remove from pending queue
      this.pendingCommands.delete(commandId);
      
      // Emit event for failed command
      this.emit('command:failed', { commandId, error });
      
      if (error instanceof Error) {
        // Reject the original promise
        pendingCommand.reject(error);
        throw error;
      }
      
      const genericError = new Error('Command execution failed');
      pendingCommand.reject(genericError);
      throw genericError;
    }
  }

  /**
   * Deny a pending command
   * @param commandId ID of the command to deny
   * @param reason Reason for denial
   */
  public denyCommand(commandId: string, reason: string = 'Command denied'): void {
    const pendingCommand = this.pendingCommands.get(commandId);
    if (!pendingCommand) {
      throw new Error(`No pending command with ID: ${commandId}`);
    }

    // Remove from pending queue
    this.pendingCommands.delete(commandId);
    
    // Emit event for denied command
    this.emit('command:denied', { commandId, reason });
    
    // Reject the original promise
    pendingCommand.reject(new Error(reason));
  }
}
